File: src/Enterprise/Java/AdapterExample.java
interface TemperatureReader {
    int readCelsius();
}

final class LegacyFahrenheitSensor {
    int readFahrenheit() {
        return 86;
    }
}

final class FahrenheitSensorAdapter implements TemperatureReader {
    private final LegacyFahrenheitSensor sensor;

    FahrenheitSensorAdapter(LegacyFahrenheitSensor sensor) {
        this.sensor = sensor;
    }

    @Override
    public int readCelsius() {
        return (sensor.readFahrenheit() - 32) * 5 / 9;
    }
}

public final class AdapterExample {
    public static void main(String[] args) {
        var legacy = new LegacyFahrenheitSensor();
        TemperatureReader reader = new FahrenheitSensorAdapter(legacy);
        System.out.printf("legacy=%dF%n", legacy.readFahrenheit());
        System.out.printf("adapted=%dC%n", reader.readCelsius());
    }
}

File: src/Enterprise/Kotlin/AdapterExample.kt
interface TemperatureReader {
    fun readCelsius(): Int
}

class LegacyFahrenheitSensor {
    fun readFahrenheit(): Int = 86
}

class FahrenheitSensorAdapter(
    private val sensor: LegacyFahrenheitSensor,
) : TemperatureReader {
    override fun readCelsius(): Int = (sensor.readFahrenheit() - 32) * 5 / 9
}

fun main() {
    val legacy = LegacyFahrenheitSensor()
    val reader: TemperatureReader = FahrenheitSensorAdapter(legacy)
    println("legacy=${legacy.readFahrenheit()}F")
    println("adapted=${reader.readCelsius()}C")
}
